### PR TITLE
Enable openPMD standard for new diagnostics

### PR DESCRIPTION
**Pull request by MaxThevenet**
This PR makes the new diagnostics openPMD-capable. This is done by adding a new class `FlushFormatOpenPMD` so that `FlushFormat` now has 2 derived classes: `FlushFormatPlotfile` and `FlushFormatOpenPMD`. The implementation in there is a copy-paste from `WarpXIO.cpp`. This lead to a few additional changes:
- openPMD format cannot write raw data (just as before). Variables `plot_F` and `plot_rho` were renamed `plot_raw_F` and `plot_raw_rho` as they are only used for for raw output in `FlushFormatPlotfile`, which wasn't clear with the previous naming.
- Minor changes were done in `WarpXOpenPMD`: (i) MultiParticleContainer is passed by reference instead of `unique_ptr` for consistency, and the name of output directory is handled in a slightly different way, to be consistent with plotfiles.

The PR was tested on `Examples/Physics_applications/uniform_plasma/inputs_3d` with
```
warpx.openpmd_int = 10
warpx.openpmd_backend = h5

diagnostics.ndiags = 1
diagnostics.diags_names = d1
d1.period = 10
d1.diag_type = Full
d1.format = openpmd
d1.openpmd_backend = h5
```
and checking the output with this kind of openPMD-viewer commands
```py
import numpy as np
from opmd_viewer import OpenPMDTimeSeries
old = OpenPMDTimeSeries('./h5/')
new = OpenPMDTimeSeries('./diags/d1h5/')
oldf = old.get_field('E','x',iteration=0)
newf = new.get_field('E','x',iteration=0)
np.all(np.equal(oldf[0],newf[0])) # check field data
newf[1].xmin == oldf[1].xmin # check field metadata
var_list=['B/x', 'E/y', 'charge', 'id', 'mass', 'x', 'y', 'z', 'ux', 'uy', 'uz', 'w']
oldp = old.get_particle(species='electrons', var_list=var_list, iteration=10) 
newp = new.get_particle(species='electrons', var_list=var_list, iteration=10)
for i in range(len(newp)):  
    print(np.amax( np.abs( (newp[i]-oldp[i]) ) ) )
```

This should be the last capability required to delete old diagnostics. While this PR doesn't have much deletion, a following PR will delete the old implementation of diagnostics.

### CODE REVIEW

File: Source/Diagnostics/FlushFormats/FlushFormatOpenPMD.H [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        bool plot_raw_fields,
+        bool plot_raw_fields_guards,
+        bool plot_raw_rho, bool plot_raw_F) const override final;
+
+    ~FlushFormatOpenPMD ();
+
+private:
+    /** Which backend to use (ADIOS, ADIOS2 or HDF5). Default depends on what is available */
+    std::string openpmd_backend {"default"};
+    bool openpmd_tspf = true; //!< one file per timestep (or one file for all steps)
```
Comment (RemiLehe) [inline Source/Diagnostics/FlushFormats/FlushFormatOpenPMD.H:35]: Should this be prefixed with `m_`?
```suggestion
    std::string m_openpmd_backend {"default"};
    bool m_openpmd_tspf = true; //!< one file per timestep (or one file for all steps)
```

File: Source/Diagnostics/FlushFormats/FlushFormatOpenPMD.cpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+namespace
+{
+    const std::string level_prefix {"Level_"};
+}
+
+FlushFormatOpenPMD::FlushFormatOpenPMD (const std::string& diag_name)
+{
+    ParmParse pp(diag_name);
+    pp.query("openpmd_backend", openpmd_backend);
+    pp.query("openpmd_tspf", openpmd_tspf);
```
Comment (RemiLehe) [inline Source/Diagnostics/FlushFormats/FlushFormatOpenPMD.cpp:18]:
```suggestion
    pp.query("openpmd_backend", m_openpmd_backend);
    pp.query("openpmd_tspf", m_openpmd_tspf);
```

File: Source/Diagnostics/FlushFormats/FlushFormatOpenPMD.cpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+using namespace amrex;
+
+namespace
+{
+    const std::string level_prefix {"Level_"};
+}
+
+FlushFormatOpenPMD::FlushFormatOpenPMD (const std::string& diag_name)
+{
+    ParmParse pp(diag_name);
```
Comment (RemiLehe) [inline Source/Diagnostics/FlushFormats/FlushFormatOpenPMD.cpp:16]:
```suggestion
    std::string openpmd_backend {"default"};
    bool openpmd_tspf = true;
    ParmParse pp(diag_name);
```

**Comment by ax3l** [comment]
Awesome, this looks great!
Besides Rémi's comments, it's potentially good to double-check with an LWFA example since the non-collective writes in it can check some corner cases (some ranks write no particles, some write multiple times, etc.). The overall logic that handles this is not changed here, but it's just good to make sure.

File: Source/Diagnostics/FlushFormats/FlushFormatOpenPMD.H (hunk 1 above)
Comment (MaxThevenet) [reply Source/Diagnostics/FlushFormats/FlushFormatOpenPMD.H:35]: Agreed.

File: Source/Diagnostics/FlushFormats/FlushFormatOpenPMD.cpp (hunk 3 above)
Comment (MaxThevenet) [reply Source/Diagnostics/FlushFormats/FlushFormatOpenPMD.cpp:16]: Thanks, I included that change.

**Comment by MaxThevenet** [comment]
@RemiLehe concerning the documentation, I thought it would be painful to also duplicate the doc and delete the old one afterwards. So the new digs are currently un-documented, and I would like to update the doc and input files at the same time as I delete the old diags. Would that work for you?

@ax3l Nah, these are little changes, shouldn't introduce any bug... huh? What is that segfault?
This test uncovered a bug introduced in PR #856, we're working on it.

Again, thanks for reviewing this PR, I will start the cleaning process once this one is merged :).

**Comment by RemiLehe** [comment]
OK, sounds good. I agree the documentation could be updated in another PR.

**Pull request merged**